test/common/http/http2/codec_impl_test.cc

```diff
@@ -947,30 +954,68 @@ TEST_P(Http2CodecImplTest, ConnectionKeepaliveJitter) {
   client_http2_options_.mutable_connection_keepalive()->mutable_timeout()->set_seconds(1);
   client_http2_options_.mutable_connection_keepalive()->mutable_interval_jitter()->set_value(10);
   /*auto timeout_timer = */ new NiceMock<Event::MockTimer>(&client_connection_.dispatcher_);
-  auto send_timer = new Event::MockTimer(&client_connection_.dispatcher_);
+  auto send_timer = new NiceMock<Event::MockTimer>(&client_connection_.dispatcher_);
 
   constexpr std::chrono::milliseconds min_expected(1000);
   constexpr std::chrono::milliseconds max_expected(1099); // 1000ms + 10%
   std::chrono::milliseconds min_observed(5000);
   std::chrono::milliseconds max_observed(0);
   EXPECT_CALL(*send_timer, enableTimer(_, _))
       .WillRepeatedly(Invoke([&](const std::chrono::milliseconds& ms, const ScopeTrackedObject*) {
+        send_timer->enabled_ = true;
         EXPECT_GE(ms, std::chrono::milliseconds(1000));
         EXPECT_LE(ms, std::chrono::milliseconds(1100));
         max_observed = std::max(max_observed, ms);
         min_observed = std::min(min_observed, ms);
       }));
   initialize();
+  ASSERT_TRUE(send_timer->enabled());
 
+  // Don't allow synchronous dispatch, it results in receiving the ping under
+  // the stack of sending it.
   for (uint64_t i = 0; i < 250; i++) {
+    client_wrapper_.dispatching_ = true;
+
     EXPECT_CALL(random_, random()).WillOnce(Return(i));
-    send_timer->callback_();
+    ASSERT_TRUE(send_timer->enabled());
+    send_timer->invokeCallback();
+
+    client_wrapper_.dispatching_ = false;
+    client_wrapper_.dispatchBufferedData().IgnoreError();
+    client_wrapper_.dispatching_ = true;
   }
 
   EXPECT_EQ(min_observed.count(), min_expected.count());
   EXPECT_EQ(max_observed.count(), max_expected.count());
 }
 
+TEST_P(Http2CodecImplTest, IdlePing) {
+  client_http2_options_.mutable_connection_keepalive()
+      ->mutable_connection_idle_interval()
+      ->set_seconds(1);
+  client_http2_options_.mutable_connection_keepalive()->mutable_timeout()->set_seconds(1);
+  client_http2_options_.mutable_connection_keepalive()->mutable_interval_jitter()->set_value(10);
+  auto timeout_timer = new NiceMock<Event::MockTimer>(&client_connection_.dispatcher_);
+
+  initialize();
+
+  // Given the initial stream is close to connection establishment, no ping is
+  // sent.
+  EXPECT_CALL(*timeout_timer, enableTimer(_, _)).Times(0);
+  TestRequestHeaderMapImpl request_headers;
+  HttpTestUtility::addDefaultHeaders(request_headers);
+  EXPECT_CALL(request_decoder_, decodeHeaders_(_, true));
+  EXPECT_TRUE(request_encoder_->encodeHeaders(request_headers, true).ok());
+
+  // Advance time past 1s. This time the ping should be sent, and the timeout
+  // alarm enabled.
+  RequestEncoder* request_encoder2 = &client_->newStream(response_decoder_);
+  client_connection_.dispatcher_.time_system_.advanceTimeAsyncImpl(std::chrono::seconds(2));
+  EXPECT_CALL(*timeout_timer, enableTimer(_, _)).Times(0);
```

Comment:
Guessing a test that verifies that we're actually sending a PING is not that easy to write? :) 

Fine with this, just curious

Reply:
yeah, there's no stats, we just tell nghttp2 to send it and get bytes on the wire.